src/constructLayoutEngine.js

```diff
@@ -63,56 +72,97 @@ export function constructLayoutEngine({
   function arrangeDomElements() {
     const path = location[resolvedRoutes.mode === "hash" ? "hash" : "pathname"];
 
-    // TODO memoize this for perf
-    const matchedRoutes = matchRoute(resolvedRoutes, path);
+    if (!path.startsWith(baseWithoutSlash)) {
+      // Base URL doesn't match, no need to recurse routes
+      return;
+    }
 
-    const initialContainer =
+    const parentContainer =
       typeof resolvedRoutes.containerEl === "string"
         ? document.querySelector(resolvedRoutes.containerEl)
         : resolvedRoutes.containerEl;
 
-    recurseRoutes(matchedRoutes.routes, initialContainer);
+    recurseRoutes({
+      path,
+      pathMatch: baseWithoutSlash,
+      routes: resolvedRoutes.routes,
+      parentContainer,
+      shouldMount: true,
+    });
+  }
+
+  function handleAppChange({ detail: { appsByNewStatus } }) {
+    appsByNewStatus.NOT_MOUNTED.concat(appsByNewStatus.NOT_LOADED).forEach(
+      (name) => {
+        const applicationElement = getApplicationElement(name);
+        if (applicationElement.isConnected) {
+          applicationElement.parentNode.removeChild(applicationElement);
+        }
+      }
+    );
   }
 }
 
 /**
+ * @typedef {{
+ * path: string,
+ * pathMatch: string,
+ * routes: Array<import('./constructRoutes').Route>,
+ * parentContainer: HTMLElement,
+ * previousSibling?: HTMLElement,
+ * shouldMount: boolean;
+ * }} DomChangeInput
  *
- * @param {Array<import('./constructRoutes').Route>} routes
- * @param {HTMLElement} parentContainer
- * @param {HTMLElement=} previousSibling
+ * We do all of this in a single recursive pass for performance, even though
+ * it makes the code a bit messier
+ *
+ * @param {DomChangeInput} input
  * @returns {HTMLElement}
  */
-function recurseRoutes(routes, parentContainer, previousSibling) {
+function recurseRoutes({
+  path,
+  pathMatch,
+  routes,
+  parentContainer,
+  previousSibling,
+  shouldMount,
+}) {
   routes.forEach((route) => {
     if (route.type === "application") {
       const applicationContainer = getContainerEl(parentContainer, route);
-      const applicationElement = getApplicationElement(route);
-
-      if (
-        previousSibling &&
-        previousSibling.parentNode === applicationContainer
-      ) {
-        // move to be immediately after previousSibling
-        previousSibling.insertAdjacentElement("afterend", applicationElement);
-      } else if (applicationElement.parentNode !== applicationContainer) {
-        // append to end of the container
-        applicationContainer.appendChild(applicationElement);
-      }
-
-      // Only use this as the reference sibling node if it's within the parent container
-      if (applicationContainer === parentContainer) {
-        previousSibling = applicationElement;
+      const applicationElement = getApplicationElement(route.name);
+
+      if (shouldMount) {
```

Comment:
This shouldMount flag is new - in the next PR I will be handling when it's falsy, but currently it is only handled when it's truthy